### PR TITLE
Add functions to autoupdate unmodified COW views

### PR DESCRIPTION
**Pull request by Yajo**
In Odoo v12+, with `website` installed, and with its COW system for view edits, when a user clicks enables or disables a view marked as `customize_show=True` by using the dropdown in the website editor, Odoo creates a COW website-specific view which is enabled or disabled as desired.

When updating, COW views are usually not updated. If they contain any logic, it can lead to template rendering problems easily.

These errors make sense on some circumstances, but if it was just because the user used the customize show feature, they are disturbing and meaningless. With these methods provided here, this problem can be reduced to only meaninful occurrences.

The 1s method will compare website-specific views with their agnostic counterparts. If bodies are equal, they will be marked.

The 2nd method takes all those marked views and resets their bodies to the new upstream, which possibly comes from the module and is updated for the new Odoo version.

@Tecnativa TT23136

### CODE REVIEW

File: openupgradelib/openupgrade.py [hunk 1]
```diff
@@ -2606,3 +2608,86 @@ def convert_to_company_dependent(
...
+       pre-migration.
+    """
+    mark_colname = mark_colname or get_legacy_name("cow_equal_to_upstream")
+    mark_identifier = sql.Identifier(mark_colname)
+    logged_query(
+        cr,
+        sql.SQL("""
+            UPDATE ir_ui_view AS specific
+            SET arch_db = agnostic.arch_db
+            FROM ir_ui_view AS agnostic
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade.py:2686]:
```suggestion
            FROM ir_ui_view AS generic
```

File: openupgradelib/openupgrade.py [hunk 2]
```diff
@@ -2606,3 +2608,86 @@ def convert_to_company_dependent(
...
+            WHERE
+                specific.{} AND
+                agnostic.website_id IS NULL AND
+                specific.key = agnostic.key
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade.py:2690]:
```suggestion
                specific.key = agnostic.key AND
                specific.type = 'qweb' AND
                specific.key IS NOT NULL AND
                generic.type = 'qweb'
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Can't be done in a step?

**Comment by Yajo** [comment]
It's not possible. 🤷🏼‍♂️ If you were to compare a specific view with its generic counterpart in post-update, maybe the generic one would have been updated, producing a false negative that would never update a view. The need to split in pre and post is technical.

**Review by pedrobaeza — APPROVED** [review]
Thanks for this!

File: openupgradelib/openupgrade.py [hunk 3]
```diff
@@ -2606,3 +2608,89 @@ def convert_to_company_dependent(
...
+    if not column_exists(cr, "ir_ui_view", mark_colname):
+        logged_query(
+            cr,
+            sql.SQL("ALTER TABLE ir_ui_view ADD COLUMN {} BOOLEAN")
+            .format(mark_identifier))
+    # Map all qweb views
+    cr.execute("""
+        SELECT id, arch_db, key, website_id
+        WHERE type == 'qweb' AND key IS NOT NULL
+    """)
```
Comment (kos94ok-3D) [inline openupgradelib/openupgrade.py:2636]: I don`t see `FROM` statement

Comment (Yajo) [reply openupgradelib/openupgrade.py:2636]: Oops... https://github.com/OCA/openupgradelib/pull/223

File: openupgradelib/openupgrade.py [hunk 4]
```diff
@@ -2606,3 +2608,89 @@ def convert_to_company_dependent(
...
+        # ever touching the view, the COW system duplicates `arch_db`
+        # preserving even the same whitespace.
+        # In case this is proven false under some circumstances, we would need
+        # to do a smarter XML comparison here.
+        if arch_db == upstream_arch:
+            equal.append(id_)
+    # Mark equal views
+    logged_query(
+        cr,
+        sql.SQL("UPDATE ir_ui_view SET {} = TRUE WHERE ID IN %s")
```
Comment (MiquelRForgeFlow) [inline openupgradelib/openupgrade.py:2661]: error in this query. see website PR

Comment (Yajo) [reply openupgradelib/openupgrade.py:2661]: See #225

**Pull request merged**